packages/hardhat/contracts/Pools.sol

```diff
@@ -0,0 +1,161 @@
+import "prb-math/contracts/PRBMath.sol";
+import "prb-math/contracts/PRBMathUD60x18.sol";
+
+import "./interfaces/IPools.sol";
+import "./interfaces/Validated.sol";
+
+contract Pools is IPools, Validated {
+    using SafeERC20 for ERC20;
+    using PRBMathUD60x18 for uint256;
+
+    // This constant represents the utilization rate at which the pool aims to obtain most competitive borrow rates.
+    uint256 public immutable optimalUtilisationRate;
+    // This constant represents the excess utilization rate above the optimal.
+    // It's always equal to 1-optimal utilization rate. Added as a constant here for gas optimizations.
+    uint256 public immutable excessUtilisationRate;
+    // Base variable borrow rate when Utilization rate = 0.
+    uint256 internal immutable baseBorrowRate;
+    // Slope of the variable interest curve when utilization rate > 0 and <= optimalUtilisationRate.
+    uint256 internal immutable slope1;
+    // Slope of the variable interest curve when utilization rate > optimalUtilisationRate.
+    uint256 internal immutable slope2;
+
+    mapping(address => Pool) private pools;
+
+    constructor (
+        uint256 _optimalUtilisationRate,
+        uint256 _baseBorrowRate,
+        uint256 _slope1,
+        uint256 _slope2
+    ) {
+        optimalUtilisationRate = _optimalUtilisationRate;
+        excessUtilisationRate = PRBMathUD60x18.SCALE - _optimalUtilisationRate;
+        baseBorrowRate = _baseBorrowRate;
+        slope1 = _slope1;
+        slope2 = _slope2;
+    }
+
+    function balance(address token) external view returns (uint256) {
+      return pools[token].balance;
+    }
+
+    function borrowed(address token) external view returns (uint256) {
+      return pools[token].borrowed;
+    }
+
+    function available(address token) external view override returns (uint qty) {
+        qty = pools[token].balance - pools[token].borrowed;
+    }
+
+    function balanceOf(address token, address owner) external view returns (uint) {
+        Pool storage pool = pools[token];
+        return pool.totalShare > 0
+        ? PRBMath.mulDiv(pool.balance, pool.shares[owner], pool.totalShare)
+        : 0;
+    }
+
+    function shareOf(address token, address owner) external view returns (uint) {
+        Pool storage pool = pools[token];
+        return pool.totalShare > 0
+        ? PRBMath.mulDiv(100, pool.shares[owner], pool.totalShare)
+        : 0;
+    }
+
+    function deposit(address token, uint256 amount) external validUAmount(amount) returns (uint256 share) {
+
+        Pool storage pool = pools[token];
+
+        share = pool.totalShare > 0
+        ? PRBMath.mulDiv(amount, pool.totalShare, pool.balance)
+        : amount * (10 ** ERC20(token).decimals());
+
+        pool.balance = pool.balance + amount;
+        pool.totalShare = pool.totalShare + share;
+        pool.shares[msg.sender] = pool.shares[msg.sender] + share;
+
+        ERC20(token).safeTransferFrom(msg.sender, address(this), amount);
```

Comment:
We should go back to use `IERC20` here, if you look in openzeppeling, there are other interfaces that `ERC20` also implements that had the extra methods you're looking for.